yt_dlp/extractor/servus.py

```diff
@@ -15,130 +13,90 @@ class ServusIE(InfoExtractor):
-                continue
-            key = attribute.get('fieldKey')
-            value = attribute.get('fieldValue')
-            if not key or not value:
-                continue
-            attrs[key] = value
-
-        title = attrs.get('title_stv') or video_id
-        alt_title = attrs.get('title')
-        description = attrs.get('long_description') or attrs.get('short_description')
-        series = attrs.get('label')
-        season = attrs.get('season')
-        episode = attrs.get('chapter')
-        duration = float_or_none(attrs.get('duration'), scale=1000)
+        season = video.get('season')
         season_number = int_or_none(self._search_regex(
             r'Season (\d+)', season or '', 'season number', default=None))
+        episode = video.get('chapter')
         episode_number = int_or_none(self._search_regex(
             r'Episode (\d+)', episode or '', 'episode number', default=None))
 
         return {
             'id': video_id,
-            'title': title,
-            'alt_title': alt_title,
-            'description': description,
-            'thumbnail': thumbnail,
-            'duration': duration,
-            'timestamp': unified_timestamp(video.get('lastPublished')),
-            'series': series,
+            'title': video.get('title'),
+            'description': self._get_description(video_id) or video.get('description'),
+            'thumbnail': video.get('poster'),
+            'duration': float_or_none(video.get('duration')),
+            'timestamp': unified_timestamp(video.get('currentSunrise')),
+            'series': season,
             'season': season,
             'season_number': season_number,
             'episode': episode,
             'episode_number': episode_number,
             'formats': formats,
+            'subtitles': subtitles,
         }
+
+    def _get_description(self, video_id):
+        info = self._download_json(
+            f'https://backend.servustv.com/wp-json/rbmh/v2/media_asset/aa_id/{video_id}?fieldset=page',
+            video_id, fatal=False)
+        return traverse_obj(info, 'stv_long_description', 'stv_short_description')
+
+    def _report_errors(self, video):
+        if 'playabilityErrors' not in video:
+            raise ExtractorError('No videoUrl, and also no information about errors')
+        for error in video.get('playabilityErrors'):
+            if error == 'FSK_BLOCKED':
+                details = video['playabilityErrorDetails']['FSK_BLOCKED']
+                if 'minEveningHour' in details:
+                    raise ExtractorError(
+                        f'Only playable from {details["minEveningHour"]}:00 to {details["maxMorningHour"]}:00',
+                        expected=True)
+            elif error == 'NOT_YET_AVAILABLE':
+                raise ExtractorError('Only available after ' + video.get('currentSunrise'), expected=True)
+            else:
+                raise ExtractorError(f'Not playable: {error}')
```

Comment:
While I like your idea, the code will still fatal for multiple reasons. If you want to proof it against that, use `traverse_obj` and `.get()` whenever possible and use a f-string for the last one: `f'Only available after {video.get("currentSunrise")}'`

Reply:
I have one doubt about the logic here. If `FSK_BLOCKED in errors and minEveningHour not in details`, the code falls into the other branches. Is this intentional?

Reply:
Good point, did not want to change the logic too much from the for loop. But I agree, it should rather raise an exception, that it is FSK blocked, but there is no information on when the video is available

Reply:
Gave it my best shot using .get and f strings, in case this was not sufficient please suggest a change. Will also just verify them with servus.tv videos

Reply:
Good thing I manually verified, found out myself that the code was not good.

Test results (WIP):

NOT_YET_AVAILABLE - works, but we should also use playabilityErrorDetails ("use information from multiple sources")
![image](https://user-images.githubusercontent.com/36509607/212550080-07e91181-6acf-46ff-be59-7a1d946995ef.png)

Reply:
Final test (using 6e96ce5):

```
C:\Users\Stefan\IdeaProjects\yt-dlp\venv\Scripts\python.exe -m yt_dlp https://www.servustv.com/unterhaltung/v/aajr8naee2h80r93kkpa/ 
[Servus] Extracting URL: https://www.servustv.com/unterhaltung/v/aajr8naee2h80r93kkpa/
[Servus] AAJR8NAEE2H80R93KKPA: Downloading video JSON
ERROR: [Servus] aajr8naee2h80r93kkpa: Only playable from 22:00 to 6:00

Process finished with exit code 1
```

```
C:\Users\Stefan\IdeaProjects\yt-dlp\venv\Scripts\python.exe -m yt_dlp https://www.servustv.com/aktuelles/v/aatc901bjsnh3gsbpa8i/ 
[Servus] Extracting URL: https://www.servustv.com/aktuelles/v/aatc901bjsnh3gsbpa8i/
[Servus] AATC901BJSNH3GSBPA8I: Downloading video JSON
ERROR: [Servus] aatc901bjsnh3gsbpa8i: Only available after 2023-01-15T21:52:59.640Z

Process finished with exit code 1
```

```
C:\Users\Stefan\IdeaProjects\yt-dlp\venv\Scripts\python.exe -m yt_dlp https://www.servustv.com/unterhaltung/v/aa9rxetmq91ef33yp4de/ 
[Servus] Extracting URL: https://www.servustv.com/unterhaltung/v/aa9rxetmq91ef33yp4de/
[Servus] AA9RXETMQ91EF33YP4DE: Downloading video JSON
[Servus] AA9RXETMQ91EF33YP4DE: Downloading m3u8 information
[Servus] AA9RXETMQ91EF33YP4DE: Downloading JSON metadata
[info] AA9RXETMQ91EF33YP4DE: Downloading 1 format(s): hls-5977+hls-audio-stereo-aac-Deutsch
[hlsnative] Downloading m3u8 manifest
WARNING: You have requested merging of multiple formats but ffmpeg is not installed. The formats won't be merged
[hlsnative] Total fragments: 1942
[download] Destination: Taktik [AA9RXETMQ91EF33YP4DE].fhls-5977.mp4
[download]   0.3% of ~   2.27GiB at    1.41MiB/s ETA 12:49 (frag 5/1942)
ERROR: Interrupted by user

Process finished with exit code 1
```

Reply:
Here is the version I cooked up. It should be safe agains errors, feel free to remove the `assert`s
```py
    def _report_errors(self, video):
        playability_errors = traverse_obj(video, ('playabilityErrors', ...))
        assert isinstance(playability_errors, list)
        if not playability_errors:
            raise ExtractorError('No videoUrl and no information about errors')

        if 'FSK_BLOCKED' in playability_errors:
            details = traverse_obj(video, ('playabilityErrorDetails', 'FSK_BLOCKED'), expected_type=dict)
            assert isinstance(details, dict) or details is None
            message = ''
            if details:
                if details.get('minEveningHour') is not None:
                    message += f' from {details["minEveningHour"]:0>2}:00'
                if details.get('maxMorningHour') is not None:
                    message += f' to {details["maxMorningHour"]:0>2}:00'
                if message and details.get('minAge') is not None:
                    message += f' (Minimum age {details["minAge"]})'

            message = (
                f'Only available{message}' if message
                else 'Blocked by FSK with unknown availability')

        elif 'NOT_YET_AVAILABLE' in playability_errors:
            available_from = traverse_obj(
                video, ('playabilityErrorDetails', 'NOT_YET_AVAILABLE', 'availableFrom'), 'currentSunrise')
            message = (
                f'Only available from {available_from}' if available_from
                else 'Video not yet available with unknown availability')

        else:
            message = f'Video unavailable: {", ".join(playability_errors)}'

        raise ExtractorError(message, expected=True)
```

Reply:
Used your code now fe85c740, I dont mind the asserts, I am assuming they help with type safety in your IDE?

Reply:
Yes, properly typing `traverse_obj` is not really possible in Python and the inferred type is inaccurate, which is why I used the `assert`s for "typing" instead.